File: leetcode/lc-us/cpp/1061.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        f.resize(123);
        for (char c = 'a'; c <= 'z'; c++) f[c] = c;
        for (int i = 0; i < s1.length(); i++) connect(s1[i], s2[i]);
        string res;
        for (char c: baseStr) {
            res.push_back(find(c));
        }
        return res;
    }

private:
    vector<char> f;

    char find(char c) {
        return c == f[c] ? c : find(f[c]);
    }

    void connect(char x, char y) {
        char fx = find(x);
        char fy = find(y);
        if (fx == fy) return;
        if (fx > fy) swap(fx, fy);
        f[fy] = fx;
    }
};

File: leetcode/lc-us/py/1061.py
class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        p = [i for i in range(26)]

        def find(x: int) -> int:
            if p[x] != x:
                p[x] = find(p[x])
            return p[x]

        for i in range(len(s1)):
            a, b = ord(s1[i]) - ord('a'), ord(s2[i]) - ord('a')
            pa, pb = find(a), find(b)
            if pa > pb:
                p[pa] = pb
            else:
                p[pb] = pa

        res = ''
        for c in baseStr:
            res += chr(find(ord(c) - ord('a')) + ord('a'))
        return res
